### ISSUE TITLE
502 if I return data via the FastCGI stderr channel

### ISSUE THREAD

**Issue by kamulos** [haproxy/haproxy#743]
## haproxy -vv
```
HA-Proxy version 2.1.7 2020/06/09 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2021.
Known bugs: http://www.haproxy.org/bugs/bugs-2.1.7.html
Running on: Linux 5.7.7-arch1-1 #1 SMP PREEMPT Wed, 01 Jul 2020 14:53:16 +0000 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wno-implicit-fallthrough -Wno-stringop-overflow -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS =

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT -PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED -REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +BACKTRACE +GETADDRINFO -OPENSSL -LUA +FUTEX +ACCEPT4 -MY_ACCEPT4 -ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=8).
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built without PCRE or PCRE2 support (using libc's regex instead)
Encrypted password support via crypt(3): yes
Built without compression support (neither USE_ZLIB nor USE_SLZ are set).
Compression algorithms supported : identity("identity")

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services : none

Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[TRACE] trace
	[COMP] compression
```

## uname -a
```
Linux xyxy 5.7.7-arch1-1 #1 SMP PREEMPT Wed, 01 Jul 2020 14:53:16 +0000 x86_64 GNU/Linux
```

## What's the configuration?

```
defaults
  mode http
  timeout connect 5000ms
  timeout client 50000ms
  timeout server 50000ms


backend cloud1
    server server0 127.0.0.1:9002 proto fcgi
    use-fcgi-app cloud-app

fcgi-app cloud-app
    docroot /
    no option keep-conn

frontend web
    bind *:80
    default_backend cloud1
```

## Steps to reproduce the behavior

Compile this program with the libfcgi:
```
include <fcgi_stdio.h>

void main(void)
{
  while (FCGI_Accept() >= 0)   {
    printf("Content-type: text/html\r\n"
           "\r\n"
           "a");
    fprintf(stderr, "b");
  }
}
```

start with `spawn-fcgi -n -p 9002 -- ./a.out`

## Actual behavior

This results in `HTTP 502 The server returned an invalid or incomplete response.` and does not return the data in the HTTP response. The stderr data is logged.

If no stderr data is sent by the program, the stdout data is in the HTTP response as expected.

## Expected behavior

In nginx the stdout data is returned with HTTP 200 and the stderr data is logged to a file.

**Comment by kamulos** [comment]
I guess the reason for this is, that there is no `fcgi_strm_handle_empty_stderr` at [mux_fcgi.c:2561](https://github.com/haproxy/haproxy/blob/master/src/mux_fcgi.c#L2561) as it is at [mux_fcgi.c:2558](https://github.com/haproxy/haproxy/blob/master/src/mux_fcgi.c#L2558) . Now [b_xfer failes](https://github.com/haproxy/haproxy/blob/master/src/mux_fcgi.c#L2367) because `fconn->drl` is `0` . The stderr stream like the stdout stream ends with an empty record ([see here](http://www.mit.edu/~yandros/doc/specs/fcgi-spec.html#S3.3))

**Comment by kamulos** [comment]
This is a quick and possibly not correct fix I created:

```
			case FCGI_STDERR:
				if (fstrm->flags & FCGI_SF_ES_RCVD)
					goto ignore_record;

				TRACE_PROTO("receiving FCGI STDERR record", FCGI_EV_RX_RECORD|FCGI_EV_RX_STDERR, fconn->conn, fstrm);
				if (fconn->drl)
					ret = fcgi_strm_handle_stderr(fconn, fstrm);
				else
					ret = fcgi_strm_handle_empty_stdout(fconn, fstrm);
				break;
```

It just reuses the `fcgi_strm_handle_empty_stdout` function. This is obviously wrong in terms of the name. But now everything seems to work fine.

**Comment by capflam** [comment]
There is indeed a bug. But we cannot fix it calling `fcgi_strm_handle_empty_stdout()` when an empyt FCGI_STDERR record is received because the `FCGI_SF_ES_RCVD` flag will be set on the fcgi stream, skipping all following FCGI_STDOUT records. Instead, it seems reasonable to skip the data processing in `fcgi_strm_handle_stderr()` if the record length is equal to 0.

I pushed the fix to the 2.3-dev
